### DISCUSSION TITLE
Would it be a good idea to expose an async as well as a sync version of the exported functions?

### DISCUSSION THREAD

**Discussion by Amatewasu** [emscripten-core/emscripten#18757] [Ideas]
I don't know how other people use emscripten but when I call a C function from JS, most of the time the function is quite "heavy" (at least a few ms). I use it a bit like a backend. So in this case I don't want to block my JS main thread.

I mean, an option to expose a synchronous version as well as an asynchronous version of the exported function could be a good idea. Thereby, in the JS code, when we want to block the JS main thread we use the synchronous version, and want we want a Promise we use the asynchronous version.

Of course, it would increase the output but I think it could be ok.

**Comment by Amatewasu** [comment]
Maybe the solution would simply to enable `-sASYNCIFY` by default. Thereby one could either call `MyWasmModule.ccall('heavy_stuff_1', [], [], [], { async: true });` or `MyWasmModule.ccall('heavy_stuff_1', [], [], [], { async: false});` depending of its need?

**Reply by sbc100** [reply]
In your example, where would `heavy_stuff_1` run if not on the main thread?   Are you perhaps imagining scheduling the work on a background thread?

**Reply by Amatewasu** [reply]
Yeah I agree, it is likely way better to call the function from a worker. It is just that I expected the wasm VM to act a bit like a worker.

**Reply by sbc100** [reply]
Yeah, wasm can't do that on its own.  Unless you run it in worker any actual work will always block them main thread for as long as it runs.

**Reply by Amatewasu** [reply]
Alright, thank you a lot for the explanation. Thereby, the feature idea is useless. Have a nice day. :)
